Tensile/Source/client/source/Reference.cpp

```diff
@@ -118,6 +118,60 @@ namespace Tensile
             throw std::runtime_error(msg.c_str());
         }
 
+        /* high precision accumulate: dist type != Accumulator or src type != Accumulator, such as H/H/H + HPA, H/H/S + HPA, B/B/S + HPA, B/S/S + HPA
+         * Ignore types: I8/I32/I32, I8x4/I32/I32
+         */
+        template <typename Inputs, typename Accumulator>
+        constexpr bool is_hpa
+            = !(std::is_same<typename Inputs::AType, Accumulator>()
+                && std::is_same<typename Inputs::BType, Accumulator>()
+                && std::is_same<typename Inputs::DType, Accumulator>())
+              && !(std::is_same<typename Inputs::AType, Int8>()
+                   && std::is_same<typename Inputs::DType, int32_t>()) //case I8/I32/I32
+              && !(std::is_same<typename Inputs::AType, Int8x4>()
+                   && std::is_same<typename Inputs::DType, int32_t>()); //case I8x4/I32/I32
+
+        /* is_hpa and (A!=B or A!=Accumlator or B!=Accumlator) */
+        template <typename A, typename B, typename Inputs, typename Accumulator>
+        constexpr bool need_transform
+            = is_hpa<
+                  Inputs,
+                  Accumulator> && !(std::is_same<A, Accumulator>() && std::is_same<B, Accumulator>() && std::is_same<A, B>());
```

Comment:
seems don't need to check  A==B.
If A==B, it means that one of condition A!=ACC and condition B!=ACC must be true.

Reply:
can we simply use this check??
need_transform = (A!=ACC or B!=ACC or A! = I8 or A!=I8x4)

Reply:
Yes, simpling the condition is possible under the current type combinations, 

Three cases need to check the need_trasnform.
- AxB
- alpha x rC
- Beta x D

So the idea here is whether these two input types are consistent or not; if not, then needs to do the type transform.
Also, we need to filter out cases when A=i8 or A=8x4

Here is the idea,
- AxB :  (A!=ACC or B!=ACC) and A!=I8 and A!=I8x4
- alpha x rC :   ((alpha!=ACC or rC!=ACC) and A!=I8   and A!=I8x4
- Beta x D : (beta!=ACC or D!=ACC )and A!=I8 and A!=I8x4

And here is the result of doing the transformation or not, the **expected need_transmform** is the same as the **result_need_transform**. 
>NoteL If the need_transform is true, the two inputs will be cast to ACC before doing the multiplication.
>For the case, B/B/S +!HPA, which beta/alpha will cast to the type B, if we want to use higher precision(type S) then still need to specialize this case.


  |   |   |   |   | expected need_transform | | | result need_transform| | |
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
A/B | C/D | Alpha/Beta | ACC | HPA | A x B | alpha x rC | Beta x C | AxB | alpha x rC | beta x D
B | B | S | B | FALSE | FALSE | TRUE | TRUE | FALSE | TRUE | TRUE
C | C | C | C | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
D | D | D | D | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
H | H | H | H | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
I | I | I | I | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
S | S | S | S | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
Z | Z | Z | Z | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
B | B | S | S | TRUE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE
B | S | S | S | TRUE | TRUE | FALSE | FALSE | TRUE | FALSE | FALSE
H | H | H | S | TRUE | TRUE | TRUE | TRUE | TRUE | TRUE | TRUE
H | H | S | S | TRUE | TRUE | FALSE | TRUE | TRUE | FALSE | TRUE
H | S | S | S | TRUE | TRUE | FALSE | FALSE | TRUE | FALSE | FALSE
I8 | I | I | I | TRUE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE
I8x4 | I | I | I | TRUE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE